src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1106 @@
+    {
+        if (is_squashed()) return false;
+        Tick now = curTick();
+
+        waitingForInstXlation.sample(now - queue_time);
+        // The event callback for when the translation completed comes through
+        // a BaseTLB::Translation object, so we pass that callback straight to
+        // the requestor's callback function by subclassing it.
+        BaseTLB::Translation* handler = new CallbackTransHandler(
+                                            this, callback_func, now, true);
+
+        tc->getITBPtr()->translateTiming(req, tc, handler, BaseTLB::Execute);
+
+        return true;
+    });
+
+    instAddrTranslationUnit.schedule();
+}
+
+void
+FlexCPU::requestInstructionData(const RequestPtr& req,
+    FetchCallback callback_func, std::function<bool()> is_squashed)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestInstructionData()\n");
+
+    // Allocate some memory for objects we're about to send through the Port
+    // system
+    PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+    pkt->allocate();
+
+    // Note: pkt is deleted in InstPort::recvTimingResp();
+
+    Tick queue_time = curTick();
+
+    instFetchUnit.addRequest([this, req, callback_func, pkt, queue_time,
+                              is_squashed]
+    {
+        if (is_squashed()) {
+            delete pkt;
+            return false;
+        }
+
+        waitingForInstData.sample(curTick() - queue_time);
+
+        //  Don't send any requests while we're waiting for recvReqRetry
+        _instPort.sendPacket(pkt);
+
+        DPRINTF(FlexCPUCoreEvent, "Sent a fetch request(pa: %#x)\n",
+                                pkt->req->getPaddr());
+
+        // We mark the control flow so that we know what to call when memory
+        // responds
+        outstandingFetches.emplace(pkt, callback_func);
+
+        return true;
+    });
+
+    instFetchUnit.schedule();
+}
+
+void
+FlexCPU::requestIssue(function<void()> callback_func,
+                                std::function<bool()> is_squashed)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestIssue()\n");
+
+    issueUnit.addRequest([callback_func, is_squashed] {
+        if (is_squashed()) return false;
+        callback_func();
+        return true;
+    });
+
+    issueUnit.schedule();
+}
+
+
+bool
+FlexCPU::requestMemRead(const RequestPtr& req, ThreadContext* tc,
+                                  StaticInstPtr inst,
+                                  weak_ptr<ExecContext> context,
+                                  Trace::InstRecord* trace_data,
+                                  MemCallback callback_func)
+{
+    DPRINTF(FlexCPUCoreEvent, "requestMemRead()\n");
+
+    // Request can be handled immediately
+    if (req->getFlags().isSet(Request::NO_ACCESS)) {
+        return false; // TODO handle now.
+    }
+
+    PacketPtr pkt = Packet::createRead(req);
+    pkt->allocate();
+
+    // Note: pkt is deleted in DataPort::recvTimingResp();
+
+    // Alpha specific?
+    if (req->isMmappedIpr()) {
```

Comment:
Double check this is still needed.

Reply:
Every CPU model still does this it seems. The actual implementation is in a separate feature commit.